core/dbt/contracts/graph/nodes.py

```diff
@@ -492,6 +502,46 @@ def depends_on_nodes(self):
     def depends_on_macros(self):
         return self.depends_on.macros
 
+    def build_contract_checksum(self):
+        # We don't need to construct the checksum if the model does not
+        # have contract enabled, because it won't be used.
+        # This needs to be executed after contract config is set
+        if self.contract is True:
+            contract_state = ""
+            # We need to sort the columns so that order doesn't matter
+            # columns is a str: ColumnInfo dictionary
+            sorted_columns = sorted(self.columns.values(), key=lambda col: col.name)
+            for column in sorted_columns:
+                contract_state += f"|{column.name}"
+                contract_state += str(column.data_type)
+            data = contract_state.encode("utf-8")
+            self.contract_checksum = hashlib.new("sha256", data).hexdigest()
+
+    def same_contract(self, old) -> bool:
+        if old.contract is False and self.contract is False:
+            # Not a change
+            return True
+        if old.contract is False and self.contract is True:
+            # A change, but not a breaking change
+            return False
+
+        breaking_change_reasons = []
+        if old.contract is True and self.contract is False:
+            # Breaking change: throw an error
+            # Note: we don't have contract_checksum for current node, so build
+            self.build_contract_checksum()
```

Comment:
Just checking my understanding: instead of building the contract checksum for every model, we only build it for models with enabled contracts or previously enabled contracts (which we detect at this point, in the state:modified check).

Reply:
That's the current state, yes. Does that feel right to you?

Reply:
Yes - this feels right! Nice way to ensure we can detect the full set of breaking changes without having to unnecessarily compute a checksum for every model, even those that with unenforced contracts. The comment is definitely helpful here though! 👍